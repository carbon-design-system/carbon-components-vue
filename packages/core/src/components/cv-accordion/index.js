import CvAccordion from './cv-accordion';
import CvAccordionItem from './cv-accordion-item';
import CvAccordionSkeleton from './cv-accordion-skeleton';

export { CvAccordion, CvAccordionItem, CvAccordionSkeleton };
