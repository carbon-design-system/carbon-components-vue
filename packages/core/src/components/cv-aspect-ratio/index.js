import CvAspectRatio from './cv-aspect-ratio';
export { CvAspectRatio };
