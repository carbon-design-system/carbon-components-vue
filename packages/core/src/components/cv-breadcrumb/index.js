import CvBreadcrumb from './cv-breadcrumb';
import CvBreadcrumbItem from './cv-breadcrumb-item';
import CvBreadcrumbSkeleton from './cv-breadcrumb-skeleton';

export { CvBreadcrumb, CvBreadcrumbItem, CvBreadcrumbSkeleton };
