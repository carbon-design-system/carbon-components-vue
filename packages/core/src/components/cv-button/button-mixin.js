import { settings as carbonSettings } from 'carbon-components';

export default {
  props: {
    icon: {
      type: [String, Object],
      default: undefined,
      validator(val) {
        if (!val || typeof val === 'string') {
          return true;
        }
        return val.render !== null;
      },
    },
    iconHref: {
      type: String,
      default: undefined,
      validator(val) {
        if (val !== undefined && process.env.NODE_ENV === 'development') {
          console.warn('CvButton: iconHref deprecated in favour of icon to be removed in future versions.');
        }
        return true;
      },
    },
    kind: {
      type: String,
      default: 'primary',
      validator: val => ['primary', 'secondary', 'tertiary', 'ghost', 'danger', 'danger--primary'].includes(val),
    },
    small: {
      type: Boolean,
      default: undefined,
      validator(val) {
        if (val !== undefined && process.env.NODE_ENV === 'development') {
          console.warn('CvButton: small deprecated in favour of size.');
        }
        return true;
      },
    },
    size: { type: String, default: undefined, validator: val => ['', 'field', 'small'].includes(val) },
  },
  computed: {
    // Bind listeners at the component level to the embedded input element and
    // add our own input listener to service the v-model. See:
    // https://vuejs.org/v2/guide/components-custom-events.html#Customizing-Component-v-model
    inputListeners() {
      return Object.assign({}, this.$listeners, {
        click: event => this.$emit('click', event),
      });
    },
    buttonClassOpts() {
      return (opts = {}) => {
        let classes = [`${carbonSettings.prefix}--btn`];

        if (opts.skeleton) {
          classes.push(`${carbonSettings.prefix}--skeleton`);
        }

        if (opts.iconOnly) {
          classes.push(`${carbonSettings.prefix}--btn--icon-only`);
        }

        if (this.kind && !opts.skeleton) {
          classes.push(`${carbonSettings.prefix}--btn--${this.kind.toLowerCase()}`);
        }

        if (this.size === 'small' || (this.size === undefined && this.small)) {
          classes.push(`${carbonSettings.prefix}--btn--sm`);
        }
        if (this.size === 'field') {
          classes.push(`${carbonSettings.prefix}--btn--field`);
        }

        return classes.join(' ');
      };
    },
  },
};
