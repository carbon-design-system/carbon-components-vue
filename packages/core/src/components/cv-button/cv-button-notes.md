# cv-button

A Vue implementation of a Carbon Components button.

http://www.carbondesignsystem.com/components/button/code

## Usage

```html
<cv-button kind="secondary" :icon="iconPath" @click="actionClick">
  Hello I am a button
</cv-button>
```

## Attributes

- kind: 'primary' (default), 'secondary', 'tertiary', 'ghost', 'danger', 'danger--ghost' or 'danger--tertiary'. Optional.
- small: (deprecated prefer size) If true smaller version of button.
- size: optional value; default, 'field', 'sm' (small), 'lg' (large), 'xl' (extra large). NOTE: At Carbon 10.23.1 icon-button treats 'lg' and 'xl' as default.
- icon: is optional. It takes an Vue Component expected to be a an icon that follows the pattern used for fill/color/stroke in Carbon Icons. It can be in the form of: a component (e.g. @carbon/icons-vue), an SVG symbol path, an SVG path, raw SVG.
- icon-href: deprecated in favour of icon attribute. Expects tring path to SVG icon..

NOTE: Recommend using icons as components or SVG symbols.
NOTE 2: Using an SVG path without an element ID will not style correctly. This is due to SVG1.1 not supporting 'use' tag without ID being specified, as a result an img tag is used.

# cv-icon-button

Rather than add extra properties that do not work with the standard button, an icon only button has been created. If in future the standard button supports the tooltip then these properties will be merged back into CvButton.

```html
<cv-icon-button
  :kind="kind"
  :size="size"
  :disabled="disabled"
  :icon="icon"
  :tip-text="tipText"
  :tip-position="tipPosition"
  :tip-alignment="tipAlignment"
/>
```

## Attributes

As per CvButton with the addition of

- selected: whether the Button is currently selected
- tip-text: String displayed in tip
- tip-position: top, left, bottom or right
- tip-alignment: start, center or end.

# cv-button-set

A wrapper for a set of cv-button components that formats them as a group

## Attributes

- stacked: Button set is arranged vertically

```html
<cv-button-set>
  <cv-button kind="primary">button 1</cv-button>
  <cv-button kind="secondary">button 2</cv-button>
  <cv-button kind="danger">button 3</cv-button>
</cv-button-set>
```
