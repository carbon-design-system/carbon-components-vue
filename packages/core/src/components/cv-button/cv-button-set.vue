<template>
  <div class="cv-button-set" :class="`${carbonPrefix}--btn-set`">
    <slot />
  </div>
</template>

<script>
import { carbonPrefixMixin } from '../../mixins';

export default {
  name: 'CvButtonSet',
  mixins: [carbonPrefixMixin],
};
</script>
