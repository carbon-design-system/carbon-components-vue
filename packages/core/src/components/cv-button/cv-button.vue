<template>
  <button class="cv-button" :class="buttonClassOpts()" v-on="inputListeners" role="button">
    <slot></slot>

    <CvSvg v-if="icon || iconHref" :svg="icon || iconHref" :class="`${carbonPrefix}--btn__icon`" />
  </button>
</template>

<script>
import buttonMixin from './button-mixin';
import { carbonPrefixMixin } from '../../mixins';
import CvSvg from '../cv-svg/_cv-svg';

export default {
  name: 'CvButton',
  mixins: [buttonMixin, carbonPrefixMixin],
  components: { CvSvg },
};
</script>
