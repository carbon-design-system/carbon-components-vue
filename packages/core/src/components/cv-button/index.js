import CvButton from './cv-button';
import CvIconButton from './cv-icon-button';
import CvButtonSkeleton from './cv-button-skeleton';
import CvButtonSet from './cv-button-set';

export { CvButton, CvIconButton, CvButtonSkeleton, CvButtonSet };
