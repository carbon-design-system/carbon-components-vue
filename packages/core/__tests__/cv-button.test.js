// import { shallowMount as shallow } from '@vue/test-utils';
import { testComponent, awaitNextTick } from './_helpers';
const { shallowMount: shallow, trigger } = awaitNextTick;
import { CvButton, CvIconButton, CvButtonSkeleton } from '@/components/cv-button';
import AddFilled16 from '@carbon/icons-vue/es/add--filled/16';
// import { settings as carbonSettings } from 'carbon-components';

// const carbonPrefix = carbonSettings.prefix;

describe('CvButton', () => {
  // ***************
  // PROP CHECKS
  // ***************

  // Deprecated props not tested
  testComponent.propsHaveDefault(CvButton, ['kind']);
  testComponent.propsAreType(CvButton, ['iconHref', 'kind', 'size'], String);
  testComponent.propsAreType(CvButton, ['icon'], [String, Object]);
  testComponent.propsHaveDefaultOfUndefined(CvButton, ['size', 'icon']);

  // ***************
  // SNAPSHOT TESTS
  // ***************

  it('Renders as expected field secondary with icon disabled', async () => {
    const propsData = { kind: 'secondary', icon: AddFilled16, 'tab-index': 2, disabled: true, size: 'field' };
    const wrapper = await shallow(CvButton, { propsData, slots: { default: 'default slot content' } });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Renders as expected small primary', async () => {
    const propsData = { kind: 'primary', size: 'small' };
    const wrapper = await shallow(CvButton, { propsData, slots: { default: 'default slot content' } });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Renders as expected default', async () => {
    const propsData = {};
    const wrapper = await shallow(CvButton, { propsData, slots: { default: 'default slot content' } });

    expect(wrapper.html()).toMatchSnapshot();
  });

  // ***************
  // FUNCTIONAL TESTS
  // ***************
  it('Raises click event when clicked', async () => {
    const wrapper = await shallow(CvButton);
    await trigger(wrapper.find('button'), 'click');
    expect(wrapper.emitted().click).toBeTruthy();
  });
});

describe('CvIconButton', () => {
  // ***************
  // PROP CHECKS
  // ***************

  // Deprecated props not tested
  testComponent.propsHaveDefault(CvIconButton, ['kind']);
  testComponent.propsAreType(
    CvIconButton,
    ['iconHref', 'kind', 'size', 'label', 'tipPosition', 'tipAlignment'],
    String
  );
  testComponent.propsAreType(CvIconButton, ['icon'], [String, Object]);
  testComponent.propsHaveDefault(CvIconButton, ['kind']);
  testComponent.propsHaveDefaultOfUndefined(CvIconButton, ['size', 'icon']);

  // ***************
  // SNAPSHOT TESTS
  // ***************

  it('Renders as expected field secondary with icon disabled', async () => {
    const propsData = { kind: 'secondary', icon: AddFilled16, 'tab-index': 2, disabled: true, size: 'field' };
    const wrapper = await shallow(CvIconButton, { propsData, slots: { default: 'default slot content' } });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Renders as expected small primary', async () => {
    const propsData = { kind: 'primary', size: 'small' };
    const wrapper = await shallow(CvIconButton, { propsData, slots: { default: 'default slot content' } });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Renders size (sm) as expected small primary', async () => {
    const propsData = { kind: 'primary', size: 'sm' };
    const wrapper = await shallow(CvIconButton, { propsData, slots: { default: 'default slot content' } });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Renders as expected lg primary', async () => {
    const propsData = { kind: 'primary', size: 'lg' };
    const wrapper = await shallow(CvIconButton, { propsData, slots: { default: 'default slot content' } });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Renders as expected lg primary', async () => {
    const propsData = { kind: 'primary', size: 'xl' };
    const wrapper = await shallow(CvIconButton, { propsData, slots: { default: 'default slot content' } });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Renders as expected default', async () => {
    const propsData = {};
    const wrapper = await shallow(CvIconButton, { propsData, slots: { default: 'default slot content' } });

    expect(wrapper.html()).toMatchSnapshot();
  });

  // ***************
  // FUNCTIONAL TESTS
  // ***************
  it('Raises click event when clicked', async () => {
    const wrapper = await shallow(CvIconButton);
    await trigger(wrapper.find('button'), 'click');
    expect(wrapper.emitted().click).toBeTruthy();
  });
});

describe('CvButtonSkeleton', () => {
  // ***************
  // PROP CHECKS
  // ***************

  // Deprecated props not tested
  testComponent.propsAreType(CvButtonSkeleton, ['size'], String);
  testComponent.propsHaveDefaultOfUndefined(CvButtonSkeleton, ['size']);

  // ***************
  // SNAPSHOT TESTS
  // ***************

  it('Renders as expected field primary', async () => {
    const propsData = { size: 'field' };
    const wrapper = await shallow(CvButtonSkeleton, { propsData, slots: { default: 'default slot content' } });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Renders as expected small primary', async () => {
    const propsData = { size: 'small' };
    const wrapper = await shallow(CvButtonSkeleton, { propsData, slots: { default: 'default slot content' } });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Renders as expected default', async () => {
    const propsData = {};
    const wrapper = await shallow(CvButtonSkeleton, { propsData, slots: { default: 'default slot content' } });

    expect(wrapper.html()).toMatchSnapshot();
  });

  // ***************
  // FUNCTIONAL TESTS
  // ***************
});
