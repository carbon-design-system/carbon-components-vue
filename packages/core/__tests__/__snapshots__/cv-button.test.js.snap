// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CvButton Renders as expected default 1`] = `
<button role="button" class="cv-button bx--btn bx--btn--primary">default slot content
  <!----></button>
`;

exports[`CvButton Renders as expected field secondary with icon disabled 1`] = `<button role="button" class="cv-button bx--btn bx--btn--secondary bx--btn--field" tab-index="2" disabled="disabled">default slot content <cvsvg-stub svg="[object Object]" class="bx--btn__icon"></cvsvg-stub></button>`;

exports[`CvButton Renders as expected small primary 1`] = `
<button role="button" class="cv-button bx--btn bx--btn--primary bx--btn--sm">default slot content
  <!----></button>
`;

exports[`CvButtonSkeleton Renders as expected default 1`] = `<button disabled="disabled" type="button" class="bx--btn bx--skeleton"></button>`;

exports[`CvButtonSkeleton Renders as expected field primary 1`] = `<button disabled="disabled" type="button" class="bx--btn bx--skeleton bx--btn--field"></button>`;

exports[`CvButtonSkeleton Renders as expected small primary 1`] = `<button disabled="disabled" type="button" class="bx--btn bx--skeleton bx--btn--sm"></button>`;

exports[`CvIconButton Renders as expected default 1`] = `
<button type="button" class="cv-button bx--btn bx--btn--icon-only bx--btn--primary bx--tooltip__trigger bx--tooltip--a11y bx--tooltip--bottom bx--tooltip--align-center"><span class="bx--assistive-text"></span>
  <!----></button>
`;

exports[`CvIconButton Renders as expected field secondary with icon disabled 1`] = `
<button type="button" class="cv-button bx--btn bx--btn--icon-only bx--btn--secondary bx--btn--field bx--tooltip__trigger bx--tooltip--a11y bx--tooltip--bottom bx--tooltip--align-center" tab-index="2" disabled="disabled"><span class="bx--assistive-text"></span>
  <cvsvg-stub svg="[object Object]" class="bx--btn__icon"></cvsvg-stub>
</button>
`;

exports[`CvIconButton Renders as expected lg primary 1`] = `
<button type="button" class="cv-button bx--btn bx--btn--icon-only bx--btn--primary bx--btn--lg bx--tooltip__trigger bx--tooltip--a11y bx--tooltip--bottom bx--tooltip--align-center"><span class="bx--assistive-text"></span>
  <!----></button>
`;

exports[`CvIconButton Renders as expected lg primary 2`] = `
<button type="button" class="cv-button bx--btn bx--btn--icon-only bx--btn--primary bx--btn--xl bx--tooltip__trigger bx--tooltip--a11y bx--tooltip--bottom bx--tooltip--align-center"><span class="bx--assistive-text"></span>
  <!----></button>
`;

exports[`CvIconButton Renders as expected small primary 1`] = `
<button type="button" class="cv-button bx--btn bx--btn--icon-only bx--btn--primary bx--btn--sm bx--tooltip__trigger bx--tooltip--a11y bx--tooltip--bottom bx--tooltip--align-center"><span class="bx--assistive-text"></span>
  <!----></button>
`;

exports[`CvIconButton Renders size (sm) as expected small primary 1`] = `
<button type="button" class="cv-button bx--btn bx--btn--icon-only bx--btn--primary bx--btn--sm bx--tooltip__trigger bx--tooltip--a11y bx--tooltip--bottom bx--tooltip--align-center"><span class="bx--assistive-text"></span>
  <!----></button>
`;
